### PR TITLE
Code Arena: Style and Gas

### PR DESCRIPTION
**Pull request by MikeHathaway**
<!---
No need to add special tag
src/ & non src/ changes you need the following (that apply):
-->
# Description of change
## High level
* Various style changes and gas optimizations
  * Preincrement and predecrement in `RewardsManager.sol`
  * [Simplify typecasting](https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/Audit_Avengers-G.md#gas-5-pre-perform-typecasting-in-curburnepoch)
  * [Remove unneeded `+=`](https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/0xnev-G.md#g-02-do-not-need-to-use--operator--can-be-used-instead)
  * [Remove unneeded local variable](https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/0xnev-Q.md#r-06-rewardsmanager_getburnepochsclaimed-can-be-refactored)
  * [Use local variable in `getPositionInfo`](https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/0xnev-G.md#g-07-multiple-accesses-of-a-mappingarray-should-use-a-local-variable-cache)
  * [Cache storage pointers](https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/JCN-G.md#cache-storage-pointer-for-positionsparams_tokenid-1)
  * [Expand natspec](https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/GG_Security-Q.md)

<!---
Add the `Status: Needs Auditor Approval` tags
CHANGES IN /SRC DIR:
- renaming (not retyping or resizing) of variables & methods
- reordering and moving of functions in files
- lite moving of functions accross files
- comments

src/ changes you need the following (that apply):
-->

# Description of bug or vulnerability and solution
* https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/0xSmartContract-G.md
* https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/0xnev-G.md
* https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/Audit_Avengers-G.md
* https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/GG_Security-Q.md
* https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/JCN-G.md

# Gas usage
## Pre Change
```
                                                                                                                                                                                                                   
| src/PositionManager.sol:PositionManager contract |                 |        |        |         |         |                                                                                                       
|--------------------------------------------------|-----------------|--------|--------|---------|---------|                                                                                                       
| Deployment Cost                                  | Deployment Size |        |        |         |         |                                                                                                       
| 3847409                                          | 19812           |        |        |         |         |                                                                                                       
| Function Name                                    | min             | avg    | median | max     | # calls |                                                                                                       
| approve(address,uint256)                         | 25184           | 25184  | 25184  | 25184   | 4       |                                                                                                       
| approve(address,uint256)(bool)                   | 25184           | 25184  | 25184  | 25184   | 50      |                                                                                                       
| getLP                                            | 6658            | 8646   | 7954   | 11748   | 200     |                                                                                                       
| getPositionIndexes                               | 1674            | 2201   | 2144   | 3790    | 441     |                                                                                                       
| getPositionIndexesFiltered                       | 8940            | 27716  | 26424  | 79674   | 705     |                                                                                                       
| isAjnaPool                                       | 955             | 6595   | 6824   | 10871   | 49      |                                                                                                       
| memorializePositions                             | 139009          | 383524 | 260038 | 1155463 | 54      |                                                                                                       
| mint                                             | 79197           | 85782  | 79197  | 94997   | 54      |                                                                                                       
| ownerOf                                          | 777             | 777    | 777    | 777     | 160     |                                                                                                       
| poolKey                                          | 807             | 807    | 807    | 807     | 54      |                                                                                                       
| transferFrom(address,address,uint256)            | 20021           | 28579  | 20021  | 41417   | 10      |
| transferFrom(address,address,uint256)(bool)      | 20021           | 27839  | 23897  | 41417   | 71      |


| src/RewardsManager.sol:RewardsManager contract |                 |        |        |         |         |
|------------------------------------------------|-----------------|--------|--------|---------|---------|
| Deployment Cost                                | Deployment Size |        |        |         |         |
| 1706462                                        | 8938            |        |        |         |         |
| Function Name                                  | min             | avg    | median | max     | # calls |
| calculateRewards                               | 36485           | 44558  | 38886  | 175150  | 348     |
| claimRewards                                   | 721             | 113995 | 89009  | 405971  | 350     |
| emergencyUnstake                               | 25030           | 26857  | 27401  | 27401   | 12      |
| getBucketStateStakeInfo                        | 904             | 4901   | 4904   | 4904    | 177337  |
| getStakeInfo                                   | 748             | 748    | 748    | 748     | 410     |
| isBucketUpdated                                | 720             | 720    | 720    | 720     | 170     |
| stake                                          | 157097          | 412757 | 262269 | 1154835 | 54      |
| unstake                                        | 564             | 149828 | 123764 | 410437  | 17      |
| updateBucketExchangeRatesAndClaim              | 4408            | 232709 | 187529 | 559370  | 49      |

```
## Post Change
```
                                                                                                                                                                                                                   
| src/PositionManager.sol:PositionManager contract |                 |        |        |         |         |                                                                                                       
|--------------------------------------------------|-----------------|--------|--------|---------|---------|                                                                                                       
| Deployment Cost                                  | Deployment Size |        |        |         |         |                                                                                                       
| 3872043                                          | 19935           |        |        |         |         |                                                                                                       
| Function Name                                    | min             | avg    | median | max     | # calls |                                                                                                       
| DOMAIN_SEPARATOR                                 | 732             | 732    | 732    | 732     | 9       |                                                                                                       
| PERMIT_TYPEHASH                                  | 426             | 426    | 426    | 426     | 8       |                                                                                                       
| approve(address,uint256)                         | 25302           | 25302  | 25302  | 25302   | 8       |                                                                                                       
| approve(address,uint256)(bool)                   | 2722            | 24579  | 25302  | 25302   | 34      |                                                                                                       
| burn                                             | 1838            | 5460   | 6209   | 11089   | 13      |                                                                                                       
| getLP                                            | 6658            | 9005   | 8260   | 36386   | 229     |                                                                                                       
| getPositionIndexes                               | 1674            | 2364   | 2614   | 3790    | 114     |                                                                                                       
| getPositionIndexesFiltered                       | 8940            | 36545  | 26424  | 79674   | 91      |                                                                                                       
| getPositionInfo                                  | 1047            | 2297   | 1047   | 5047    | 16      |                                                                                                       
| isAjnaPool                                       | 955             | 6473   | 6824   | 10871   | 32      |                                                                                                       
| isIndexInPosition                                | 1060            | 1623   | 1060   | 3060    | 71      |                                                                                                       
| isPositionBucketBankrupt                         | 6928            | 7944   | 8148   | 9716    | 18      |                                                                                                       
| memorializePositions                             | 6459            | 79067  | 73744  | 1154607 | 3071    |                                                                                                       
| mint                                             | 10529           | 92961  | 97764  | 102764  | 76      |                                                                                                       
| moveLiquidity                                    | 5261            | 211862 | 180605 | 689708  | 20      |                                                                                                       
| nonces                                           | 841             | 1817   | 1746   | 2841    | 8       |                                                                                                       
| ownerOf                                          | 836             | 837    | 836    | 882     | 150     |                                                                                                       
| permit                                           | 1290            | 23250  | 17585  | 44417   | 9       |                                                                                                       
| poolKey                                          | 807             | 807    | 807    | 807     | 36      |                                                                                                       
| redeemPositions                                  | 2855            | 48183  | 59992  | 158094  | 29      |                                                                                                       
| safeTransferFrom                                 | 24282           | 36677  | 41715  | 42115   | 7       |                                                                                                       
| tokenURI                                         | 3157            | 468651 | 468651 | 934146  | 2       |                                                                                                       
| transferFrom(address,address,uint256)            | 20623           | 34839  | 42068  | 42068   | 9       |                                                                                                       
| transferFrom(address,address,uint256)(bool)      | 20623           | 29739  | 24548  | 42068   | 57      |                                                                                                       
                                                                                                                                                                                                                   
                                                                                                                                                                                                                   
| src/RewardsManager.sol:RewardsManager contract |                 |        |        |         |         |                                                                                                         
|------------------------------------------------|-----------------|--------|--------|---------|---------|                                                                                                         
| Deployment Cost                                | Deployment Size |        |        |         |         |                                                                                                         
| 1705862                                        | 8935            |        |        |         |         |                                                                                                         
| Function Name                                  | min             | avg    | median | max     | # calls |                                                                                                         
| calculateRewards                               | 36162           | 60877  | 40425  | 167020  | 41      |                                                                                                         
| claimRewards                                   | 721             | 106715 | 108438 | 397840  | 43      |                                                                                                         
| emergencyUnstake                               | 25632           | 27459  | 28003  | 28003   | 12      |                                                                                                         
| getBucketStateStakeInfo                        | 904             | 4901   | 4904   | 4904    | 177337  |                                                                                                         
| getStakeInfo                                   | 748             | 748    | 748    | 748     | 83      |                                                                                                         
| isBucketUpdated                                | 720             | 720    | 720    | 720     | 170     |                                                                                                         
| stake                                          | 157678          | 501645 | 518512 | 1154247 | 34      |                                                                                                         
| unstake                                        | 564             | 149746 | 124203 | 408876  | 17      |                                                                                                         
| updateBucketExchangeRatesAndClaim              | 4408            | 255735 | 225222 | 556834  | 32      |                                                                                                         
                                                                                                             
```

### CODE REVIEW

File: src/RewardsManager.sol [hunk 1]
```diff
@@ -157,13 +157,13 @@ contract RewardsManager is IRewardsManager {
         stakeInfo.owner    = msg.sender;
         stakeInfo.ajnaPool = ajnaPool;
 
-        uint256 curBurnEpoch = IPool(ajnaPool).currentBurnEpoch();
+        uint96 curBurnEpoch = uint96(IPool(ajnaPool).currentBurnEpoch());
```
Comment (grandizzy) [inline src/RewardsManager.sol:165]: should this use a safe cast?

**Review by grandizzy — COMMENTED** [review]
looks good, just a comment re cast of current burn epoch

File: src/RewardsManager.sol (hunk 1 above)
Comment (MikeHathaway) [reply src/RewardsManager.sol:165]: Yes, probably. That being said, I don't think it's possible to ever overflow as currentBurnEpoch increases monotonically from 1...

Comment (grandizzy) [reply src/RewardsManager.sol:165]: 👍

**Review by grandizzy — APPROVED** [review]

File: src/RewardsManager.sol [hunk 2]
```diff
@@ -738,7 +736,7 @@ contract RewardsManager is IRewardsManager {
                 uint256 burnFactor = Maths.wmul(totalBurned_, bucketDeposit);
 
                 // calculate rewards earned for updating bucket exchange rate 
-                rewards_ += interestEarned_ == 0 ? 0 : Maths.wdiv(
+                rewards_ = interestEarned_ == 0 ? 0 : Maths.wdiv(
```
Comment (EdNoepel) [inline src/RewardsManager.sol:749]: Was confused this went from increment to assignment without formula change.  Then realized there's no loop, and that it was always incrementing from 0.

**Review by EdNoepel — APPROVED** [review]
LGTM.  Would be nice if _post change_ gas was filled out though.

File: src/RewardsManager.sol (hunk 2 above)
Comment (grandizzy) [reply src/RewardsManager.sol:749]: This one got me too

File: src/PositionManager.sol [hunk 3]
```diff
@@ -121,6 +121,8 @@ contract PositionManager is PermitERC721, IPositionManager, Multicall, Reentranc
         // revert if the token id is not minted for given pool address
         if (pool_ != poolKey[tokenId_]) revert WrongPool();
 
+        // TODO: verify that the pool is a valid ajna pool: https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/Bason-Q.md#-l-01--not-verifying-that-params_pool-is-valid-pool-address----------------------------------------------
+
```
Comment (MikeHathaway) [inline src/PositionManager.sol:114]: This check seems to be handled by the `WrongPool` revert: https://github.com/code-423n4/2023-05-ajna-findings/blob/main/data/Bason-Q.md#-l-01--not-verifying-that-params_pool-is-valid-pool-address----------------------------------------------

File: src/RewardsManager.sol (hunk 1 above)
Comment (MikeHathaway) [reply src/RewardsManager.sol:165]: Looking into this more - we don't currently import `SafeCast` so would have to add it as well. Still open to it

**Review by prat-gpt — APPROVED** [review]
LGTM!

**Review by EdNoepel — APPROVED** [review]
Reviewed new commits; still looks good.

**Pull request merged**